### ISSUE TITLE
This is awesome, but does it not work with standard materials?

### ISSUE THREAD

**Issue by haydeng21** [Fyrestar/ShaderMaterialExtend#2]
I stumbled upon your code here and thought my life was saved, but it appears it doesn't include capabilities to extend a standard material? If so, is it a three.js limitation?

I've been struggling trying to load in a draco gltf to replace it's meshStandardMaterial with a ShaderMaterial that copies the frag shader, vertex shader, and uniforms from shaderLib["standard"]. I get an error saying `cannot read property toArray of undefined` when mesh in injected into scene with ShaderMaterial.  When I copy lambert the error changes to an undefined property `direction`. Copying the basic material works correctly.

I'm relatively new to three.js and very new to shaders so any guidance would be really appreciated. My end goal is to extend the standard shader to be able to lerp between two textures.

**Comment by haydeng21** [comment]
Oh wow, on second look there is MeshStandardMaterial! Let me give this a try

**Comment by haydeng21** [comment]
So I got your example version of this to work on my `wheels` gltf scene using the patchShader like so:

```
const customShader = new THREE.MeshStandardMaterial({
          onBeforeCompile: function(shader) {
              THREE.patchShader(shader, {
      
                  header: 'uniform vec3 tint;',
      
                  fragment: {
                      '#include <fog_fragment>': 'gl_FragColor.rgb *= tint;'
                  },
      
                  vertex: {
                      'project_vertex': {
                          '@vec4 mvPosition = modelViewMatrix * vec4( transformed, 1.0 );':
                              'vec4 mvPosition = modelViewMatrix * vec4( transformed * 0.25, 1.0 );'
                      }
                  },
      
                  uniforms: {
                    tint: new THREE.Color('orange')
                  }
      
              });
      
          }
      });

 wheels.children[0].material = customShader;
```

However, this same implementation with the extend function doesn't work and throws the following errors:
```
index.js:1375 THREE.ShaderMaterial.extend: no mapping for material class "Gr" found
aframe-v1.0.4.min.js:95 Uncaught TypeError: Cannot set property 'value' of undefined
```

```
const customShader = THREE.ShaderMaterial.extend(wheels.children[0].material, {
    
          header: 'uniform vec3 tint;',

          fragment: {
              '#include <fog_fragment>': 'gl_FragColor.rgb *= tint;'
          },

          vertex: {
              'project_vertex': {
                  '@vec4 mvPosition = modelViewMatrix * vec4( transformed, 1.0 );':
                      'vec4 mvPosition = modelViewMatrix * vec4( transformed * 0.25, 1.0 );'
              }
          },

          uniforms: {
            tint: new THREE.Color('orange')
          }

      });
```

**Comment by Fyrestar** [comment]
This seems so be caused by the minified THREE. I've updated the code now which avoids this issue and adds other new features and improvements. See the readme for updates but your code should work now, you can pass a material class constructor as well as a instance of a built-in material and shader material. If there are still issues let me know.

**Comment by haydeng21** [comment]
Okay it's working now! At least there are no errors anymore. 

My new issue is that the values from the material I'm passing in (the wheel's gltf material) are not being cloned and passed onto customShader. An example is the metalness value, which is 0 on the wheels material but when I extend it the metalness value gets set as 0.5, the THREE.js default. 
```

 const customShader = THREE.extendMaterial(wheels.children[0].material, {
        // Will be prepended to vertex and fragment code
        header: 'uniform sampler2D mapTwo; \n uniform float fac; \n',


        // Insert code lines by hinting at a existing
        vertex: {
        },


        fragment: {
          'map_fragment': {
            "@#include <map_fragment>" : '#ifdef USE_MAP vec4 texelColor = texture2D( map, vUv ); vec4 texelColorTwo = texture2D( mapTwo, vUv ); texelColor = mapTexelToLinear( mix( texelColor, texelColorTwo, fac)); diffuseColor *= texelColor; #endif \n'
          }
        },


        // Properties to apply to the new THREE.ShaderMaterial
        material: {
          defines: {
            STANDARD: '',
            USE_MAP: true,
            USE_UV: true,
          },
          extensions: {
            derivatives: true,
          },
          lights: true
        },


        // Uniforms (will be applied to existing or added) as value or uniform object
        uniforms: {
          map: {
            value: null,
          },
          mapTwo: {
              value: null,
          },
          fac: {
              value: 0.0
          },
        }
      });
```

**Comment by haydeng21** [comment]
Figured out I can pass in uniform values (metalness: 0) that match the wheels material but is there a better way than using the uniforms?

**Comment by Fyrestar** [comment]
I'll check out later, i'm also currently making another adaption handling all cases of uniform inheritation/override and creation from in-built materials more specific, especially of when to clone and when to share.

In your case it should already assign these values to the uniforms, i'll check it later ^^

**Comment by haydeng21** [comment]
Thank you for the quick responses and fixing that 0 issue! I took what you said into consideration and have found a simple solution using onBeforeCompile.

```
var wheelsMat = wheels.children[0].material;

wheelsMat.userData.map_from = { value: wheelsMat.map };
wheelsMat.userData.fac = { value: 0.0 };

wheelsMat.onBeforeCompile = (shader) => {

  THREE.patchShader(shader, {

    // adds to top of frag and vert shader
    header: `uniform sampler2D map_from; \n uniform float fac; \n`,

    // replaces map_fragment with replacement
    fragment: {
      'map_fragment' : {
        replacement:
          `
          #ifdef USE_MAP
              vec4 tCol = texture2D( map, vUv );
              vec4 tColFrom = texture2D( map_from, vUv );

              tCol = sRGBToLinear( 
                mix( tCol, tColFrom, fac)
              );

              diffuseColor *= tCol;
          #endif
          `
      }
    },

    // define uniforms that reference userData values
    uniforms: {
      map_from: wheelsMat.userData.map_from,
      fac: wheelsMat.userData.fac,
    }
  });
}

wheelsMat.userData.lerp = true; // used to check if this mat has lerp abilities
```

It's great because it adds onto the original material from the gltf and then during runtime I can update wheelsMat.userData values and influence the uniforms I added.

Also, I wanted to replace all of the map_fragment chunk and didn't see an easy way to do it with the syntax you supported (please correct me if I'm wrong here) so I swapped out one line of code in ShaderMaterialExtend.js. If there's a way to do it with your build I'd prefer that, so please let me know.

```
// chunk = chunk.replace( '#include <' + name + '>', applyPatches( THREE.ShaderChunk[ name ], value ) );
chunk = chunk.replace( '#include <' + name + '>', value.replacement );
```


On another note, I did try to use onBeforeCompile with your `MaterialPlugins.js` but I was unable to succeed with the patch. I don't need to take this route but thought I'd let you know for feedback purposes.

Anytime I tried to use a patch (like https://codepen.io/Fyrestar/pen/ymjqMm) and assign it in the onBeforeCompile array I got the following error:

![image](https://user-images.githubusercontent.com/18315056/91221467-4c486900-e6e3-11ea-9775-40b01ba86bb0.png)

I don't have that code anymore or I'd share it but I do remember trying to pass in empty objects into the patch and it still threw the same error.

**Comment by Fyrestar** [comment]
> It's great because it adds onto the original material from the gltf and then during runtime I can update wheelsMat.userData values and influence the uniforms I added.

Notice that these uniforms are only global and not individually when only using onBeforeCompile (can't be different for multiple meshes), but if you don't need that kind of control onBeforeCompile would be enough.

Regarding the error: which version of THREE did you use? i updated the code yesterday to fix compatibility with latest THREE versions.

**Comment by haydeng21** [comment]
I did notice that wheelsMat.uniforms was undefined, but could you elaborate on what you mean by global? Am I able to access the global store of uniforms after onBeforeCompile? I have implemented this same oBeforeCompile code to a few other mesh and their mats and updating map_from is working as expected. Are you saying that if I needed to use a different uniform, lets just say newMap, on a separate mesh/material it wouldn't work?

I'm using the three.js version that is shipped with aframe. It appears to be 0.111.6. 
![image](https://user-images.githubusercontent.com/18315056/91312523-f4a90c80-e779-11ea-84ec-0f3121658530.png)

**Comment by Fyrestar** [comment]
> Do you mean every mesh/material will now have the map_from and fac values in userData?

Yes it will only use those uniforms you provided at the onBeforeCompile callback, this is why i made the plugin in order to access and change the uniform values before every render call, with the freedom of where/how you want to store the parameters for them.

I'm not sure about your case, but assuming your 4 wheels would all fade from one texture to another this will be fine, if you could change the texture of each wheel individually this wouldn't work anymore, they all will use the uniforms of the first that got compiled/rendered.

**Comment by haydeng21** [comment]
Okay this is making more sense now. 

> I'm not sure about your case, but assuming your 4 wheels would all fade from one texture to another this will be fine, if you want to change the texture of each wheel individually this wouldn't work anymore

Is that the case because those wheels would share the same material? Because I have other mesh in the scene that are a separate material (using same map_from and fac logic though) that I'm able to lerp the textures independently of the wheels. 

Also, any update on that three.js error?

**Comment by Fyrestar** [comment]
It means you could attempt to clone the wheel materials to change each individually as you would expect with regular materials, though the uniforms are only passed once, so they will either all share one or only the first will have them.

I'll check regarding the error, it's a bit confusing with THREE as there was a kinda double structure with shader, program program etc.

**Comment by haydeng21** [comment]
Okay I understand, thanks again.

I should be updating three relatively soon so if that fixes it I’ll comment again
